### PR TITLE
KBSON-4

### PR DESCRIPTION
**Pull request by rozza**
I separated out the work into three commits:

1. Adding the Corpus Test code
    Doesn't run any actual tests, yet but imports the json files and a way to be able to read them across platform
2. Adding BsonInput / Output
    Ported from the Java driver - small simplifications as just using `ByteArray` types.  
    The corpus tests now run the hex based tests
3. Adding Json serialization 
    Uses kotlinx.serialization only and handles converting BsonValues to Json and back.
    Tried to DRY the serializers up using generics but it confuses the codegen compliation stage.
    Adding Support for data classes should be simple using custom serializers for Json primitive types
    Adding data classes to bson support should also be simple by adding a custom BsonEncoder/Decoder and piping values
    to the BsonInput/Output classes in the Serializers.

### CODE REVIEW

File: src/commonMain/kotlin/org/kbson/BsonDecimal128.kt [hunk 1]
```diff
@@ -32,7 +32,7 @@ import org.kbson.internal.Decimal128
  *
  * @property value the Decimal128 value
  */
-public class BsonDecimal128 private constructor(private val value: Decimal128) : BsonValue() {
+public class BsonDecimal128 private constructor(public val value: Decimal128) : BsonValue() {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/BsonDecimal128.kt]: Just so we don't forget: Decimal128 is exposed here as a public field, even though it's in an internal package.

File: src/commonMain/kotlin/org/kbson/BsonDocument.kt [hunk 2]
```diff
@@ -55,6 +61,18 @@ public class BsonDocument(initial: Map<String, BsonValue> = LinkedHashMap()) :
...
     public constructor(vararg pairs: Pair<String, BsonValue>) : this(pairs.toMap())
 
+    /**
+     * Returns the Bson bytes for this document
+     *
+     * @return the raw bson bytes
+     */
+    public fun toByteArray(): ByteArray {
+        return BsonBinaryWriter().use { w ->
+            w.pipeDocument(this)
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/BsonDocument.kt]: Ah, so we don't actually need a Codec. Clever!

File: src/commonMain/kotlin/org/kbson/BsonType.kt [hunk 3]
```diff
@@ -20,7 +20,7 @@ package org.kbson
  *
  * @property value the int value of this BSON type.
  */
-public enum class BsonType(public val value: Int) {
+public enum class BsonType(public val value: Byte) {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/BsonType.kt]: Consider making value a `UByte` and then you can just add `u` suffix to each of the below constants.  This seems to more clearly express the intention that the range is from 0 to 255

File: src/commonMain/kotlin/org/kbson/BsonValue.kt [hunk 4]
```diff
@@ -381,6 +383,15 @@ public sealed class BsonValue {
         return this as BsonUndefined
     }
 
+    /**
+     * Return an extended json representation of this BsonValue
+     *
+     * @return the extended json representation of this BsonValue
+     */
+    public fun toJson(): String {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/BsonValue.kt:391]: Very fancy. :)

File: src/jvmTest/kotlin/org/kbson/BsonValueApiTest.kt [hunk 5]
```diff
@@ -33,7 +33,7 @@ class BsonValueApiTest {
         val kBsonConstructors = getConstructorParams(kBsonClass)
         assertEquals(bsonConstructors, kBsonConstructors)
 
-        val bsonMethods = getMethodNames(bsonClass, listOf("parse", "toJson", "asBsonReader"))
+        val bsonMethods = getMethodNames(bsonClass, listOf("parse", "asBsonReader"))
```
Comment (jyemin) [inline src/jvmTest/kotlin/org/kbson/BsonValueApiTest.kt:36]: Remind me why these tests are necessary.  I know we have similar ones for Scala driver to ensure the API matches Java driver, but why here?

File: src/commonTest/kotlin/org/kbson/BsonDocumentTest.kt [hunk 6]
```diff
@@ -326,4 +326,32 @@ class BsonDocumentTest {
         assertEquals(expectedDocument, BsonDocument(pairs.map { BsonElement(it.first, it.second) }))
         assertEquals(expectedDocument, BsonDocument(*pairs.toTypedArray()))
     }
+
+    @Test
```
Comment (jyemin) [inline src/commonTest/kotlin/org/kbson/BsonDocumentTest.kt:330]: Add `shouldRoundTripToFromString`?

File: src/commonMain/kotlin/org/kbson/internal/Closeable.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+package org.kbson.internal
+
+/**
+ * A Closeable is a source or destination of data that can be closed.
+ *
+ * The close method is invoked to release resources that the object is holding.
+ */
+public interface Closeable {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/internal/Closeable.kt:23]: I'm curious how Kotlin survives without this, and if it does, why do we need it?

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+import org.kbson.BsonRegularExpression
+import org.kbson.BsonString
+import org.kbson.BsonSymbol
+import org.kbson.BsonTimestamp
+import org.kbson.BsonType
+import org.kbson.BsonUndefined
+import org.kbson.internal.validateOperation
+
+/** Abstract base class for BsonReader implementations. */
+@Suppress("TooManyFunctions")
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt:41]: Maybe we should bump the max?  This is getting suppressed all over the place.

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+            }
+            BsonType.UNDEFINED -> {
+                reader.readUndefined()
+                writeUndefined()
+            }
+            else -> throw IllegalArgumentException("unhandled BSON type: " + reader.currentBsonType)
+        }
+    }
+
+    public fun pipeDocument(value: BsonDocument) {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt]: If we add a `BsonDocumentReader` class then then wouldn't be necessary.  It's all internal so it's doesn't matter much, but it's different than Java driver.

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt [hunk 10]
```diff
@@ -0,0 +1,799 @@
...
+     *
+     * @param bsonValue the BsonValue
+     * @return the Json String
+     */
+    public fun toJson(bsonValue: BsonValue): String {
+        return extendedJson.encodeToString(BsonValueSerializer, bsonValue)
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:112]: I'm not sure how big the risk is of opting in to experimental API.  It it worth it?

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt [hunk 11]
```diff
@@ -0,0 +1,799 @@
...
+
+    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("BsonDocument") {}
+
+    @Suppress("ComplexMethod")
+    override fun serialize(encoder: Encoder, value: BsonValue) {
+        when (encoder) {
+            is JsonEncoder -> {
+                when (value.bsonType) {
+                    BsonType.ARRAY -> ListSerializer(BsonValueSerializer).serialize(encoder, value.asArray())
+                    BsonType.DOCUMENT -> BsonDocumentSerializer.serialize(encoder, value.asDocument())
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:124]: Can this be just:

`MapSerializer(BsonDocumentKeySerializer, BsonValueSerializer).serialize(encoder, value.asDocument())`

similar to what's above for `ARRAY`?

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt [hunk 12]
```diff
@@ -0,0 +1,799 @@
...
+            else -> throw SerializationException("Unknown decoder type: $decoder")
+        }
+    }
+}
+
+private object BsonDocumentKeySerializer : KSerializer<String> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("BsonDocumentKey", PrimitiveKind.STRING)
+
+    override fun deserialize(decoder: Decoder): String {
+        TODO("Not yet implemented")
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/serialization/Serializers.kt]: I'm not clear how we round trip keys without implementing this.  Can you explain?

File: src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+import kotlinx.serialization.json.Json
+import org.kbson.BsonDecimal128
+import org.kbson.BsonDocument
+import org.kbson.BsonSerializationException
+import org.kbson.BsonType
+import org.kbson.BsonValue
+import org.kbson.internal.Assertions.assertThrows
+import org.kbson.internal.HexUtils
+
+class ArrayTest : CorpusTest("array.json")
```
Comment (jyemin) [inline src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt:37]: Why does it have to be this way?  :) I guess there is no way to iterate all the resources in the directory, as in Java?

I couldn't figure out how to run any of the tests in IDEA  just by right clicking on something.

File: src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kbson.internal
+
+import kotlin.experimental.and
+
+/** Base64 utils to convert to and from a Base64 encoded string and ByteArray */
+@Suppress("MagicNumber")
+public object Base64Utils {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt]: Can all these public classes in `org.kbson.internal` be given the `internal` modifier instead `public`, and if not, why not?

**Review by jyemin — CHANGES_REQUESTED** [review]
Initial feedback.

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt [hunk 15]
```diff
@@ -0,0 +1,799 @@
...
+ */
+public val BsonSerializersModule: SerializersModule = SerializersModule {
+    contextual(BsonValue::class, BsonValueSerializer)
+}
+
+
+/**
+ * The Bson companion object
+ */
+public object Bson {
```
Comment (jyemin) [inline src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:87]: Are you planning to add support for generating and parsing relaxed extended JSON?  That seems to be implied by the scope goals.

File: src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+@Suppress("UnnecessaryAbstractClass")
+abstract class CorpusTest(filename: String) {
+
+    @Serializable
+    data class ValidBson(
+        val description: String,
+        @SerialName("canonical_bson") val canonicalBsonHex: String,
+        @SerialName("canonical_extjson") val canonicalJson: String,
+        @SerialName("converted_bson") val convertedBsonHex: String? = null,
+        @SerialName("degenerate_bson") val degenerateBsonHex: String? = null,
```
Comment (jyemin) [inline src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt:108]: What about `degenerate_extjson`?

File: src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+     * @param byteArray the input ByteArray
+     * @return the Base64 string representation
+     * @see [toByteArray] to decode the string back to a byte array
+     */
+    public fun toBase64String(byteArray: ByteArray): String {
+        val output = mutableListOf<Byte>()
+        var padding = 0
+        var position = 0
+        while (position < byteArray.size) {
+            var b = byteArray[position].toInt() and 0xFF shl 16 and 0xFFFFFF
```
Comment (cmelchior) [inline src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt:42]: Kotlin really messed up bit-wise operators 🙈

File: src/commonMain/kotlin/org/kbson/internal/Closeable.kt (hunk 7 above)
Comment (cmelchior) [reply src/commonMain/kotlin/org/kbson/internal/Closeable.kt:23]: It is a weird omission. I agree. It is being tracked here https://youtrack.jetbrains.com/issue/KT-31066/Add-Closeable-use-to-common-stdlib for Kotlin. All other libraries are also forced to implement their own variant right now 😢.

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+public abstract class AbstractBsonReader : BsonReader {
```
Comment (cmelchior) [inline src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt]: Does this class need to be public? Couldn't it be `internal`? Same with the rest of these classes.

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 10 above)
Comment (cmelchior) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:112]: All serializer API's are experimental right now: https://kotlinlang.org/docs/serialization.html#formats, so there is really no way around it.

It also just means that the API's might change/not be backwards compatible rather than they have bugs. Since Multiplatform support in general is alpha I think this is okay.

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt [hunk 19]
```diff
@@ -0,0 +1,799 @@
...
+     * @return the Json String
+     */
+    public fun toJson(bsonValue: BsonValue): String {
+        return extendedJson.encodeToString(BsonValueSerializer, bsonValue)
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = BsonValue::class)
+private object BsonValueSerializer : KSerializer<BsonValue> {
```
Comment (cmelchior) [inline src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:114]: This looks really promising 💯

**Review by cmelchior — COMMENTED** [review]
Since this is a massive PR I'm still trying to grasp the details, but it looks really promising. I left some minor comments for now.

File: src/commonMain/kotlin/org/kbson/BsonType.kt (hunk 3 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/BsonType.kt]: I think MinKey breaks it - 0xFF.toByte() = -1

File: src/commonMain/kotlin/org/kbson/BsonDecimal128.kt (hunk 1 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/BsonDecimal128.kt]: Good catch - made it internal

File: src/jvmTest/kotlin/org/kbson/BsonValueApiTest.kt (hunk 5 above)
Comment (rozza) [reply src/jvmTest/kotlin/org/kbson/BsonValueApiTest.kt:36]: Follows the 3rd goal from the scope:

> API should be Kotlin idiomatic, but whenever possible we should attempt to maintain API compatibility with the Java library.

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt (hunk 8 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt:41]: Good idea will add in another ticket.

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt (hunk 9 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt]: It is the same as the Java driver: See https://github.com/mongodb/mongo-java-driver/blob/r4.7.1/bson/src/main/org/bson/AbstractBsonWriter.java#L894-L901

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 11 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:124]: It could be but I still need a `BsonDocumentSerializer` for the scope field in Javascript with scope - so chose to reuse it here.

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 12 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt]: I've removed the todo - so this could be used if required.  Decoding we use JsonElements and that handles the decoding of the keys.  See `BsonValueSerializer.decoder`.

File: src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt (hunk 13 above)
Comment (rozza) [reply src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt:37]: I doesn't have to be this way - but there is no resource handling in KMP. So a plugin had to be used and a full path had to be provided to the file.  There is no KMP built in file class that lists all files in a directory either.

There is also no Parameterized tests either. You can right click on the classes and get them to run individually or the directory to run all tests in the suite.

tldr; It was a painful process to get even this working :)

File: src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt (hunk 14 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt]: Done

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 15 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:87]: No plans to at this time. As the goals only talk about passing extended Json strings and we only discussed using Strict encoding:

> Support serialization between BsonDocument and JSON, as defined by [Extended JSON 2.2.0 specification](https://github.com/mongodb/specifications/blob/master/source/extended-json.rst).
> The Realm Kotlin SDK needs a way to send BSON across the JNI boundary to Realm Core. The most straightforward way seems to be if BSON documents can be serialized to Extended JSON strings.

If there is a future need for different types of serialization then future refactorings will be needed.

File: src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt (hunk 16 above)
Comment (rozza) [reply src/commonTest/kotlin/org/kbson/corpus/CorpusTests.kt:108]: Will add :)

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt (hunk 18 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonReader.kt]: Done :)

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 15 above)
Comment (jyemin) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:87]: Both Relaxed and Strict modes are defined in the extended JSON specification, but if we only need Strict that's fine for now.  I don't recall discussing it though. @cmelchior does this work for you?

File: src/commonMain/kotlin/org/kbson/BsonType.kt (hunk 3 above)
Comment (jyemin) [reply src/commonMain/kotlin/org/kbson/BsonType.kt]: Not sure what you mean.  0xFF is 255 when it's an unsigned byte, which I think is what we want.  Or leave it an int and then we don't have to worry about it going negative for MinKey.

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt (hunk 9 above)
Comment (jyemin) [reply src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt]: Ah, but that method is private.  The public one is `void pipe(final BsonReader reader)`

File: src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt (hunk 17 above)
Comment (jyemin) [reply src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt:42]: What you don't like `shl`?

File: src/commonTest/kotlin/org/kbson/BsonDocumentTest.kt (hunk 6 above)
Comment (jyemin) [reply src/commonTest/kotlin/org/kbson/BsonDocumentTest.kt:330]: Ping on this comment.

**Review by jyemin — COMMENTED** [review]
Updated a few remaining comment threads.

File: src/commonMain/kotlin/org/kbson/BsonType.kt (hunk 3 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/BsonType.kt]: The reason I chose BsonType(value: Byte) is because when reading from bson or writing to bson its needed as a byte value. So I saved having to cast from int to byte by just using Byte. I have changed to UByte but now I have to convert to byte throughout the code.

File: src/commonTest/kotlin/org/kbson/BsonDocumentTest.kt (hunk 6 above)
Comment (rozza) [reply src/commonTest/kotlin/org/kbson/BsonDocumentTest.kt:330]: Done on line 356: `assertEquals(expectedDocument, BsonDocument(expectedDocument.toJson()))`

File: src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt (hunk 9 above)
Comment (rozza) [reply src/commonMain/kotlin/org/kbson/internal/io/AbstractBsonWriter.kt]: Gotcha, its a bit more complex but I like the symmetry of having a `BsonDocumentReader` - so removed the now unused pipe(BsonValue) methods in favor of just using the new reader.

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 15 above)
Comment (jyemin) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:87]: @cmelchior we still need an answer here, but let's not hold up this PR on a decision.  If we need relaxed JSON we can implement in a subsequent PR.

**Review by jyemin — APPROVED** [review]
Deferring relaxed extended JSON decision to a subsequent ticket/PR.

LGTM!

File: src/commonMain/kotlin/org/kbson/serialization/Serializers.kt (hunk 15 above)
Comment (cmelchior) [reply src/commonMain/kotlin/org/kbson/serialization/Serializers.kt:87]: Since we are only using extended JSON as a debugging mechanism and as an internal protocol. Just having strict support should be fine 👍

File: src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt (hunk 17 above)
Comment (cmelchior) [reply src/commonMain/kotlin/org/kbson/internal/Base64Utils.kt:42]: Guess, it is just habit, but no 😄

**Review by cmelchior — APPROVED** [review]

**Comment by rozza** [comment]
Manually rebased and merged :)
